qiskit/circuit/library/data_preparation/state_preparation.py

```diff
@@ -414,3 +418,107 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
             circuit.append(CXGate(), [msb, lsb])
 
         return circuit
+
+
+class UniformSuperpositionGate(Gate):
+    r"""
+    Class that implements a uniform superposition state, using n qubits,
+    following the Shukla-Vedula algorithm [SV24].
+
+    This gate is used to create the uniform superposition state
+    $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $ when it acts on an input
+    state $\ket{0}$ (i.e. $\ket{0...0}$). The effect on other input states is
+    deterministic but undefined.
+
+    .. note::
+    The Shukla-Vedula algorithm [SV24] provides an efficient approach for
+    creation of a uniform superposition state of the form,
+    $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $. It needs only
+    $O(\log_2 (M))$ qubits and $O(\log_2 (M))$ gates, hence providing a
+    significant improvement, in terms of reduced resources and complexity,
+    compared to the use of StatePreparation gate for preparation of uniform
+    superposition state.
+
+    **References:**
+    [SV24]
+    A. Shukla and P. Vedula, ``An efficient quantum algorithm for preparation
+    of uniform quantum superposition states,`` Quantum Information Processing,
+    23(38): pp. 1-32 (2024).
+    """
+
+    def __init__(
+        self,
+        num_superpos_states: int = 2,
+        num_qubits: Optional[int] = None,
```

Comment:
With only Python 3.7+ support now we can use the `|` operand for typehints 🙂 
```suggestion
        num_qubits: int | None = None,
```